base/inference.jl

```diff
@@ -1433,6 +1499,17 @@ function abstract_call(f::ANY, fargs, argtypes::Vector{Any}, vtypes::VarTable, s
 
     if istopfunction(tm, f, :promote_type) || istopfunction(tm, f, :typejoin)
         return Type
+    elseif length(argtypes) == 2 && istopfunction(tm, f, :typename)
+        t = argtypes[2]
+        if isa(t, Const) || isType(t)
+            val = isa(t, Const) ? t.val : t.parameters[1]
+            try
+                return Const(typename(val))
+            catch
```

Comment:
In a helper function, but yes. You have to repeat most of it to figure out why it failed anyways (TypeVars, Unions of TypeVars, Vararg uncertainty, etc.)